// @nefty/ual-renderer/dist/AuthMocks/BaseMockAuthenticator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseMockAuthenticator = void 0;
const universal_authenticator_library_1 = require("universal-authenticator-library");
class BaseMockAuthenticator extends universal_authenticator_library_1.Authenticator {
    init() {
        return Promise.resolve();
    }
    login() {
        return Promise.resolve([{}]);
    }
    getStyle() {
        return {
            text: 'Base Mock',
            textColor: 'black',
            background: 'background-color: green'
        };
    }
    getError() {
        return null;
    }
    getOnboardingLink() {
        return 'http://localhost';
    }
    isErrored() {
        return false;
    }
    isLoading() {
        return false;
    }
    reset() {
        console.info('reset base mock authenticator');
    }
    shouldRender() {
        return true;
    }
    shouldAutoLogin() {
        return false;
    }
    shouldRequestAccountName() {
        return Promise.resolve(false);
    }
    logout() {
        return Promise.resolve();
    }
    requiresGetKeyConfirmation() {
        return false;
    }
    getName() {
        return 'authenticator';
    }
}
exports.BaseMockAuthenticator = BaseMockAuthenticator;
