// @nefty/ual-renderer/dist/UALJsAbstractBaseComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UALJsAbstractBaseComponent = void 0;
class UALJsAbstractBaseComponent {
    /**
     * Creates component container and attaches generated dom to it
     */
    constructor(options) {
        this.options = options;
        // attach our component to our internal dom
        this.dom = document.createElement('div');
        this.dom.appendChild(this.generateDom());
        // attach our styles to our dom
        this.styleElement = document.createElement('style');
        this.styleElement.innerHTML = this.generateStyles();
        this.dom.appendChild(this.styleElement);
    }
    /**
     * Shows the component
     */
    show() {
        this.dom.style.display = 'block';
    }
    /**
     * Hides the component
     */
    hide() {
        this.dom.style.display = 'none';
    }
    /**
     * Attaches the dom of the component to the provided parent
     * element
     *
     * @param parent element the component dom should attach to
     */
    attach(parent) {
        parent.appendChild(this.dom);
    }
    /**
     * Helper method to return the parent of the component
     */
    getComponentElement() {
        return this.dom;
    }
}
exports.UALJsAbstractBaseComponent = UALJsAbstractBaseComponent;
