// @nefty/ual-renderer/dist/UALJs.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.UALJs = void 0;
const universal_authenticator_library_1 = require("universal-authenticator-library");
const UALJsDom_1 = require("./UALJsDom");
/**
 * Plain JS implementation for UAL Interaction with UI
 */
class UALJs extends universal_authenticator_library_1.UAL {
    /**
     *
     * @param userCallbackHandler Called with the array of users after a successful authenticator selection
     * @param chains Array of Chains the application wants to support
     * @param appName Name of the application
     * @param authenticators List of authenticators this app supports
     * @param renderConfig Optional UI rendering configuration for environments not using login
     */
    constructor(userCallbackHandler, chains, appName, authenticators, renderConfig) {
        super(chains, appName, authenticators);
        this.isAutologin = false;
        this.accountNameInputValue = '';
        if (renderConfig) {
            this.renderConfig = renderConfig;
        }
        this.userCallbackHandler = userCallbackHandler;
        this.loginUser = this.loginUser.bind(this);
    }
    /**
     * Initializes UAL: If a renderConfig was provided and no autologin authenticator
     * is returned it will render the Auth Button and relevant DOM elements.
     *
     */
    init() {
        const authenticators = this.getAuthenticators();
        // perform this check first, if we're autologging in we don't render a dom
        if (!!authenticators.autoLoginAuthenticator) {
            this.isAutologin = true;
            this.loginUser(authenticators.autoLoginAuthenticator);
            this.activeAuthenticator = authenticators.autoLoginAuthenticator;
        }
        else {
            // check for existing session and resume if possible
            this.attemptSessionLogin(authenticators.availableAuthenticators);
            if (!this.renderConfig) {
                throw new Error('Render Configuration is required when no auto login authenticator is provided');
            }
            const { containerElement, buttonStyleOverride = false, } = this.renderConfig;
            this.dom = new UALJsDom_1.UALJsDom(this.loginUser, authenticators.availableAuthenticators, containerElement, buttonStyleOverride);
            this.dom.generateUIDom();
        }
    }
    /**
     * Attempts to resume a users session if they previously logged in
     *
     * @param authenticators Available authenticators for login
     */
    attemptSessionLogin(authenticators) {
        const sessionExpiration = localStorage.getItem(UALJs.SESSION_EXPIRATION_KEY) || null;
        if (sessionExpiration) {
            // clear session if it has expired and continue
            if (new Date(sessionExpiration) <= new Date()) {
                localStorage.clear();
            }
            else {
                const authenticatorName = localStorage.getItem(UALJs.SESSION_AUTHENTICATOR_KEY);
                const sessionAuthenticator = authenticators.find((authenticator) => authenticator.getName() === authenticatorName);
                const accountName = localStorage.getItem(UALJs.SESSION_ACCOUNT_NAME_KEY) || undefined;
                this.loginUser(sessionAuthenticator, accountName);
            }
        }
    }
    /**
     * App developer can call this directly with the preferred authenticator or render a
     * UI to let the user select their authenticator
     *
     * @param authenticator Authenticator chosen for login
     * @param accountName Account name (optional) of the user logging in
     */
    loginUser(authenticator, accountName) {
        return __awaiter(this, void 0, void 0, function* () {
            let users;
            // set the active authenticator so we can use it in logout
            this.activeAuthenticator = authenticator;
            const invalidateSeconds = this.activeAuthenticator.shouldInvalidateAfter();
            const invalidateAt = new Date();
            invalidateAt.setSeconds(invalidateAt.getSeconds() + invalidateSeconds);
            localStorage.setItem(UALJs.SESSION_EXPIRATION_KEY, invalidateAt.toString());
            localStorage.setItem(UALJs.SESSION_AUTHENTICATOR_KEY, authenticator.getName());
            try {
                yield this.waitForAuthenticatorToLoad(authenticator);
                if (accountName) {
                    users = yield authenticator.login(accountName);
                    localStorage.setItem(UALJs.SESSION_ACCOUNT_NAME_KEY, accountName);
                }
                else {
                    users = yield authenticator.login();
                }
                // send our users back
                this.userCallbackHandler(users);
            }
            catch (e) {
                console.error('Error', e);
                console.error('Error cause', e.cause ? e.cause : '');
                this.clearStorageKeys();
                throw e;
            }
            // reset our modal state if we're not autologged in (no dom is rendered for autologin)
            if (!this.isAutologin) {
                this.dom.reset();
            }
        });
    }
    waitForAuthenticatorToLoad(authenticator) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve) => {
                if (!authenticator.isLoading()) {
                    resolve();
                    return;
                }
                const authenticatorIsLoadingCheck = setInterval(() => {
                    if (!authenticator.isLoading()) {
                        clearInterval(authenticatorIsLoadingCheck);
                        resolve();
                    }
                }, UALJs.AUTHENTICATOR_LOADING_INTERVAL);
            });
        });
    }
    /**
     * Clears the session data for the logged in user
     */
    logoutUser() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.activeAuthenticator) {
                throw Error('No active authenticator defined, did you login before attempting to logout?');
            }
            this.activeAuthenticator.logout();
            this.clearStorageKeys();
        });
    }
    clearStorageKeys() {
        // clear out our storage keys
        localStorage.removeItem(UALJs.SESSION_EXPIRATION_KEY);
        localStorage.removeItem(UALJs.SESSION_AUTHENTICATOR_KEY);
        localStorage.removeItem(UALJs.SESSION_ACCOUNT_NAME_KEY);
    }
}
exports.UALJs = UALJs;
UALJs.SESSION_EXPIRATION_KEY = 'ual-session-expiration';
UALJs.SESSION_AUTHENTICATOR_KEY = 'ual-session-authenticator';
UALJs.SESSION_ACCOUNT_NAME_KEY = 'ual-session-account-name';
UALJs.AUTHENTICATOR_LOADING_INTERVAL = 250;
