// @nefty/ual-renderer/dist/components/DownloadAuthenticatorModal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DownloadAuthenticatorModal = void 0;
const UALJsAbstractBaseComponent_1 = require("../UALJsAbstractBaseComponent");
class DownloadAuthenticatorModal extends UALJsAbstractBaseComponent_1.UALJsAbstractBaseComponent {
    /**
     *
     * @param options { onGoBack, onClose } // goback and close callbacks
     */
    constructor(options) {
        super(options);
    }
    close() {
        // only hide and call close if we aren't already visible
        if (this.dom.style.display !== 'none') {
            this.hide();
            this.options.onClose();
        }
    }
    /**
     * Show the download modal and provide a download button for the provided authenticator
     *
     * @param authenticator Authenticator this modal represents
     */
    open(authenticator) {
        const { text, textColor, background } = authenticator.getStyle();
        const downloadUri = authenticator.getOnboardingLink();
        const contentDescription = this.dom.querySelector('.ual-modal-content-description');
        contentDescription.textContent = `Install ${text} to Continue`;
        const contentTextBox = this.dom.querySelector('.ual-modal-text-box');
        contentTextBox.textContent = `To get started with ${text}, install the app at the link below.`;
        contentDescription.style.color = textColor;
        const getDownloadLink = this.dom.querySelector('.ual-leave-and-install');
        getDownloadLink.href = downloadUri;
        const modalElement = this.dom.querySelector('#ual-modal-get-authenticator');
        modalElement.style.background = background;
        modalElement.style.color = textColor;
        this.show();
    }
    goBack() {
        this.options.onGoBack();
    }
    generateStyles() {
        return `
      .ual-go-back {
        color: white;
        font-size: 1rem;
        cursor: pointer;
        text-align: center;
      }

      a.ual-leave-and-install {
        width: 260px;
        padding: 11px;
        margin: 4em auto;
        display: block;
        border-radius: 5px;
        font-size: 1em;
        background: rgba(255, 255, 255, 0.2);
        color: inherit;
        text-align: center;
        text-decoration: none;
      }
    `;
    }
    /**
     * Generates the Modal DOM, binds go-back, and close handlers
     */
    generateDom() {
        const component = document.createElement('div');
        component.innerHTML = `
      <div id="ual-modal-get-authenticator">
        <span class="ual-modal-close">&times;</span>
        <p class="ual-modal-content-description"></p>
        <p class="ual-modal-text-box"></p>
        <a class="ual-leave-and-install"
          href="#"
          target="_blank"
          rel="noopener noreferrer"
        >Leave and Install</a>
        <div class="ual-go-back">Go Back</div>
      </div>`;
        component.querySelector('.ual-modal-close').addEventListener('click', () => {
            this.close();
        });
        component.querySelector('.ual-go-back').addEventListener('click', () => {
            this.options.onGoBack();
        });
        return component;
    }
}
exports.DownloadAuthenticatorModal = DownloadAuthenticatorModal;
